Add a new 'close_tree' plugin to close entire trees by closing the pa…

redstreet opened the pull request:
Add a new 'close_tree' plugin to close entire trees by closing the parent.

See https://groups.google.com/g/beancount/c/c0Fm4hrxpUQ

dnicolodi commented on beancount/plugins/close_tree.py:
```diff
@@ -0,0 +1,98 @@
+
+    if DEBUG:
+        elapsed_time = time.time() - start_time
+        print("Close account tree [{:.2f}s]: {} close entries added.".format(elapsed_time, close_count))
```
From this I infer that you are worried about the execution time of this plugin. If this is the case, maybe storing the accounts tree in a flat list is not the best choice. You could use a tree data structure. Also iterating the list of entries six times is not optimal. If there are no performance issues, I would delete the debug code, as it is anyhow not discoverable or accessible by the users and there are better profiling tools.

dnicolodi commented on beancount/plugins/close_tree.py:
```diff
@@ -0,0 +1,98 @@
+
+    # Above, auto_accounts inserts an Open for 'Expenses:Non-Retirement:Auto:Fit:Gas' on 2021-06-17
+    # since it appears in a close directive. But that's not what we want. The semantic of the close
+    # when u used with this plugin is: "close all children, and close the parent if it was opened
```
Spurious `u`.

blais replied:
+1

dnicolodi commented on beancount/plugins/close_tree.py:
```diff
@@ -0,0 +1,98 @@
+
+    # plugin "beancount.plugins.auto_accounts"
+    # 2021-06-17 close Expenses:Non-Retirement:Auto:Fit
+    # plugin "beancount.plugins.close_tree"
```
I would put both `plugin` directives at the beginning.

blais replied:
+1

dnicolodi commented on beancount/plugins/close_tree.py:
```diff
@@ -0,0 +1,98 @@
+    # Above, auto_accounts inserts an Open for 'Expenses:Non-Retirement:Auto:Fit:Gas' on 2021-06-17
+    # since it appears in a close directive. But that's not what we want. The semantic of the close
+    # when u used with this plugin is: "close all children, and close the parent if it was opened
+    # earlier. Below, we filter out entries to correctly handle this scenario:
```
I had a very hard time to follow what you are doing here. I had to unpack the different steps. Starting with this ledger:
```
2019-01-01 * "Transaction"
  Expenses:Non-Retirement:Auto:Fit:Insurance      10 USD
  Expenses:Non-Retirement:Auto:Fit:Gas            20 USD
  Assets:Transfer
2021-06-17 close Expenses:Non-Retirement:Auto:Fit
```
the `bencount.plugins.auto_accounts` adds the `open` directives:
```
2019-01-01 open Assets:Transfer
2019-01-01 open Expenses:Non-Retirement:Auto:Fit:Insurance
2019-01-01 open Expenses:Non-Retirement:Auto:Fit:Gas
2019-01-01 * "Transaction"
  Expenses:Non-Retirement:Auto:Fit:Insurance      10 USD
  Expenses:Non-Retirement:Auto:Fit:Gas            20 USD
  Assets:Transfer
2021-06-17 open Expenses:Non-Retirement:Auto:Fit
2021-06-17 close Expenses:Non-Retirement:Auto:Fit
```
Note that this this generates an error, because you have the open and the close directive on the same day, and this is not currently allowed. Let's ignore this for the moment. The `beancount.plugins.close_tree` adds the missing `close` directives based on the `open` directives present in the ledger:
```
2019-01-01 open Assets:Transfer
2019-01-01 open Expenses:Non-Retirement:Auto:Fit:Insurance
2019-01-01 open Expenses:Non-Retirement:Auto:Fit:Gas
2019-01-01 * "Transaction"
  Expenses:Non-Retirement:Auto:Fit:Insurance      10 USD
  Expenses:Non-Retirement:Auto:Fit:Gas            20 USD
  Assets:Transfer
2021-06-17 open Expenses:Non-Retirement:Auto:Fit
2021-06-17 close Expenses:Non-Retirement:Auto:Fit
2021-06-17 close Expenses:Non-Retirement:Auto:Fit:Insurance
2021-06-17 close Expenses:Non-Retirement:Auto:Fit:Gas
```
What you want to do is to remove the pair of `open` and `close` directives that appear on the same day. But waht if there is also a transaction to the relative account on the same day? You would need to keep the `open` and `close` directives. I think that the error message for `open` and `close` on the same day is not correct: Beancount puts `open` directives at the beginning of day and `close` directives at the end of day, and transactions in between. Thus I don't see a problem with `open` and `close` on the same day.

redstreet replied:
1. Fixed Beancount's error message.
2. Do we need to remove the pair, if there wasn't a transaction found to the account in question on that date? We would save the potentially expensive cost of checking if a removal is needed, if we instead always inserted the `open`/`close` pair (and sometimes ended up with this additional pair).

blais commented on beancount/plugins/close_tree.py:
```diff
@@ -0,0 +1,98 @@
+
+    for entry in entries:
+        if isinstance(entry,  Close):
+            subaccounts = [a for a in opens if entry.account + ':' in a and a not in closes]
```
Use a set for speed

redstreet replied:
Changed.

blais commented on beancount/plugins/close_tree.py:
```diff
@@ -0,0 +1,98 @@
+
+    for entry in entries:
+        if isinstance(entry,  Close):
+            subaccounts = [a for a in opens if entry.account + ':' in a and a not in closes]
```
a.startswith(entry.account+ ":") or use a regexp

blais commented on beancount/plugins/close_tree.py:
```diff
@@ -0,0 +1,98 @@
+    for entry in entries:
+        if isinstance(entry,  Close):
+            subaccounts = [a for a in opens if entry.account + ':' in a and a not in closes]
+            for s in subaccounts:
```
Don't use single letter variables; rename to "account" or "sub" or something.

blais commented on beancount/plugins/close_tree.py:
```diff
@@ -0,0 +1,98 @@
+    errors = []
+
+    opens = [e.account for e in entries if isinstance(e, Open)]
+    closes = [e.account for e in entries if isinstance(e, Close)]
```
If you wanted to be more precise about closing really just the subaccounts that are open at the time of closing, you should keep track of the tree of open accounts as you process the directives and close the subaccounts that are open then. You'd also have to handle subsequent closing directives colliding with your inserted ones. But I think that would be the most correct algorithm.

blais replied:
To be more clear what I mean is that it's possible for accounts to be opened, closed and then reopened (I think, I forget a bit).

redstreet replied:
```
2014-01-01 open Assets:XBank
2015-01-01 close Assets:XBank

2016-01-01 open Assets:XBank
2017-01-01 close Assets:XBank   
```

This doesn't seem to be supported (in v2, at least). I get:
```
Duplicate open directive for Assets:XBank
Duplicate close directive for Assets:XBank
```

Do we still need to address this case? Is there a plan to support it in v3 or in the future?

redstreet replied:
Ah, found the [documentation](https://beancount.github.io/docs/beancount_language_syntax.html#directives):
> _At the moment, once an account is closed, you cannot reopen it after that date._

@blais, do you still see this to be a point of concern?

dnicolodi commented on beancount/ops/validation.py:
```diff
@@ -87,7 +87,7 @@ def validate_open_close(entries, unused_options_map):
                 try:
                     open_entry = open_map[entry.account]
-                    if entry.date <= open_entry.date:
+                    if entry.date < open_entry.date:
```
This should be a separate commit with a commit message that explain what is being changed and why. That said, I haven't checked if there is a reason for not allowing `open` and `close` directives on the same day. @blais do you remember anything about this?

blais replied:
I don't remember.
Il'l review now.

blais replied:
It does make sense to me that we should allow entries on the same day as the open though. Accepting.

dnicolodi commented on beancount/plugins/close_tree.py:
````diff
@@ -0,0 +1,64 @@
+plugin "beancount.plugins.auto_accounts"
+plugin "beancount.plugins.close_tree"
+```
+"""
````
Documentation strings do not use Markdown syntax. If anything it should be reST.

dnicolodi commented on beancount/plugins/close_tree.py:
```diff
@@ -0,0 +1,64 @@
+      entries: A list of directives. We're interested only in the Open/Close instances.
+      unused_options_map: A parser options dict.
+    Returns:
+      A tuple of entries and errors. """
```
The closing `"""` should go on a line by themselves.

dnicolodi commented on beancount/plugins/close_tree_test.py:
```diff
@@ -0,0 +1,192 @@
+        actual, _ = close_tree.close_tree(entries, {})
+        for a, e in zip(s(actual), s(expected)):
+            self.assertEqual(a.date, e.date)
+            self.assertEqual(a.account, e.account)
```
Here and everywhere else, is there a reason why `self.assertEqual(actual, expected)` does not work? Also, why is there the need to sort the actual and expected entries, can't the expected entries be in the right order?

redstreet replied:
1. Doesn't work because metadata is not the same
2. Sorting the expected entries as the actual would have been sorted makes it unintuitive IMO.
```
            2019-01-01 open Expenses:Non-Retirement:Auto:Fit:Gas
            2019-01-01 open Expenses:Non-Retirement:Auto:Fit:Insurance
            ....
            2021-06-17 open Expenses:Non-Retirement:Auto:Fit
            2021-06-17 close Expenses:Non-Retirement:Auto:Fit:Insurance
            2021-06-17 close Expenses:Non-Retirement:Auto:Fit:Gas
            2021-06-17 close Expenses:Non-Retirement:Auto:Fit
```
The above is neither alphabetical, nor consistent (ordering of opens and closes), nor does the open/close pair intuitively appear next to each other. In addition, changes to the plugin's algorithm might require all expected cases to be reordered, which is messy.

dnicolodi replied:
Please have a look, for example, at how the tests for the `auto_accounts.py` plugin are structured and follow the same pattern.

redstreet committed:
Fix: allow open and close directives on the same day

Previously, open and close directives on the same day caused an error.
However, Beancount considers opens to have occured on the beginning of a
day, and a close directive to have occurred at the end of a day.

This makes most sense, and was likely the original intention, as can be
evidenced by the fact that one can currently issue a transaction on the
same day as either a close, or an open, without errors. If so, issuing
an open and a close on the same day should also be valid. This fix
enables that.

See discussion here:
https://github.com/beancount/beancount/pull/751

redstreet committed:
Add a new 'close_tree' plugin to close entire trees by closing the parent.
